### PR TITLE
Add more CI

### PR DESCRIPTION
**Pull request by dcolligan**
- run mvn commands from python tests
- throw error on mvn warnings
- add test to compile and check _protocol_definitions.py file

This build will fail since #292 has not been resolved.  I will rebase this on top of master after that branch is merged in.  After this commit is merged into master, no branch should be merged into master that does not pass the tests.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Looks good @dcolligan, thanks.

I wonder if we need to do the third part of this: do we gain much extra by generating a `_protocoldefinitions.py`? It's going to be a bit of a maintenance headache if we need to keep two copies of this schema generation code in sync.

**Comment by pgrosu** [comment]
@dcolligan, @jeromekelleher  The following will fix it by making the changes to [lines 92 to 100 in process_schemas.py](https://github.com/dcolligan/schemas/blob/063cb7ce4c3522292d8c7a98cd8bee27c2d01cec/tests/process_schemas.py#L92-100) and should make it more Avro-compatible - though we should update it for unlimited sized `union`s:

```
                elif isinstance(field.type, avro.schema.UnionSchema):
                    t0 = field.type.schemas[0]
                    t1 = field.type.schemas[1]
                    if (isinstance(t0, avro.schema.PrimitiveSchema) and
                            t0.type == "null"):
                        if isinstance(t1, avro.schema.RecordSchema):
                            ret.append((field.name, t1.name))
                    elif (isinstance(t1, avro.schema.PrimitiveSchema) and
                            t1.type == "null"):
                        if isinstance(t0, avro.schema.RecordSchema):
                            ret.append((field.name, t0.name))
                    else:
                        raise Exception("Schema union assumptions violated")

```
Below is the output that returns without errors:

```
(testenv)[schemas]$ flake8 scripts tests
(testenv)[schemas]$ nosetests -v tests
testMaven (test_maven.TestMaven) ... ok
testProcessSchemas (test_protocol.TestProcessSchemas) ... ok

----------------------------------------------------------------------
Ran 2 tests in 34.729s

OK
(testenv)[schemas]$
```

Hope it helps,
Paul

**Comment by diekhans** [comment]
@pgrosu please make proposed code changes as pull requests against the appropriate repository.

**Comment by fnothaft** [comment]
@diekhans I think @pgrosu is suggesting an improvement to the code added in this PR. @pgrosu you can PR your changes against @dcolligan's branch.

**Comment by dcolligan** [comment]
It's a bit ambiguous which repo that PR should be made in, since I basically just copied the `process_schemas.py` code from the server repo to this repo.  So that code's real home is in server.  On the other hand, I'm checking a modified version in here and the code does have to do with validating schemas... I'd say, for now, make the PR against `server/ga4gh/scripts/process_schemas.py`.

@jeromekelleher yes, that last part is necessary.  If you think that is a headache, compare that to the headache of trying to generate the schemas when you're trying to work on something else in the server repo and finding out that an invalid `_protocol_definitions.py` file is spit out.

I think all of this discussion is getting at the question of where the code that compiles `_protocol_definitions.py` should be since, yeah, having two copies of that code does suck.  I propose we migrate any code that does schema compiling/validation/processing/etc. into schemas, but only check in `_protocol_definitions.py` into the server repo.  The server's `process_schemas.py` (or whatever it ends up being called) will just invoke code in the schema repo to generate the  `_protocol_definitions.py` file.  Thoughts?

**Comment by pgrosu** [comment]
@diekhans, @fnothaft, @dcolligan  Sorry to reply so late, since I tried several things but I seem to have some Github issues when trying to perform the pull request.  Regarding the updated code, it would be just a quick fix with something more extensive in the future.  So I performed the following steps, but it says that I've been blocked from @dcolligan's repository.  Did I do something incorrectly in the following steps?

1) First I cloned my repository of `schemas`:
```
git clone https://github.com/pgrosu/schemas.git
```

2) Then I entered the directory with:

```
cd schemas
```

3) Then I added a remote:

```
git remote add dcolligan https://github.com/dcolligan/schemas.git
```

4) Then I listed the branches:

```
C:\git_ga4gh\schemas> git branch -r
  dcolligan/more_ci
  origin/AddFragmentToReads
  origin/HEAD -> origin/master
  origin/IncludeRawDataLinks
  origin/master
```

5) Next I fetched the `more_ci` branch:

```
C:\git_ga4gh\schemas> git fetch dcolligan more_ci
remote: Counting objects: 166, done.
remote: Compressing objects: 100% (58/58), done.
remote: Total 166 (delta 45), reused 9 (delta 9), pack-reused 98Receiving object

Resolving deltas:  11% (6/51)
Resolving deltas: 100% (51/51), done.
From https://github.com/dcolligan/schemas
 * branch            more_ci    -> FETCH_HEAD
 * [new branch]      more_ci    -> dcolligan/more_ci
```

6) Then I performed `checkout`:

```
C:\git_ga4gh\schemas> git checkout --track dcolligan/more_ci
Branch more_ci set up to track remote branch more_ci from dcolligan.
Switched to a new branch 'more_ci'
```

7) Next I added the changes to the `process_schemas.py` file.

8) Then I checked the status:

```
C:\git_ga4gh\schemas>git status
On branch more_ci
Your branch is up-to-date with 'dcolligan/more_ci'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   tests/process_schemas.py

no changes added to commit (use "git add" and/or "git commit -a")
```

9) Next I performed the `commit` of the changes:

```
C:\git_ga4gh\schemas>git commit -am "Update schema validation for union type"
[more_ci 9b67b24] Update schema validation for union type
 1 file changed, 4 insertions(+)

```

10) Then I pushed the changes:

```
C:\git_ga4gh\schemas>git push origin more_ci
Username for 'https://github.com': pgrosu
Password for 'https://pgrosu@github.com':
Counting objects: 223, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (71/71), done.
Writing objects: 100% (170/170), 169.47 KiB | 0 bytes/s, done.
Total 170 (delta 68), reused 149 (delta 51)
To https://github.com/pgrosu/schemas.git
 * [new branch]      more_ci -> more_ci
```

11) Then initiated a pull request that looked like this:

pull-request

12) But after clicking the "Create pull request" button I got the following message:

refused

Any ideas of what I should try next?

Thanks,
Paul

**Comment by jeromekelleher** [comment]
@dcolligan OK, agreed. This is all rather tricky though, as we're creating a lot of coupling between the two repos. There's a lot of stuff in there which really only does affect us in the server, which would be very confusing to an outsider coming in fresh.

What if we just verified the assumed structure in this script (calling it `verify_schemas.py`), and left our process_schemas as-is in the schemas repo? There is some duplication of effort here, but hopefully `check_schemas.py` will not need much updating afterwards (except to add more checks, keeping tabs on naming conventions, etc). We are then free to evolve `process_schemas.py` as we see fit in the server repo.

**Comment by dcolligan** [comment]
@jeromekelleher   I think "creating" this coupling is fine since that coupling already exists.  The server repo already crucially depends on the schemas repo.  We are just checking if that coupling is not broken at the source.

It seems like we have three choices:
1) having 2 similar `process_schemas` files in both repos 
2) only having a `process_schemas` file in the schemas repo, and having a script in the server repo that is a front end to that functionality
3) your plan, which I don't really understand since verification and processing involve the same steps.  Also I think it's safe to assume that how we process and check the schemas will evolve over time.

I prefer number 2

**Comment by jeromekelleher** [comment]
I think it would be a mistake to have code that we depend on  in the server repo (other than the schema itself, obviously) end up in here, because (a) it doesn't belong here; (b) it will make updating the schema processing code unnecessarily difficult (PRs can move *very* slowly on this repo). 

Anyway, they _are_ different tasks. One is to generate the Python code that we use, and the other is to check that the schema has some properties that we all agree are important. The checking code should evolve over time to include a lot more stuff. An example would be checking for trailing whitespace: we can easily build this in to a schema checker, but it has no purpose in a code generator.

**Comment by pgrosu** [comment]
@jeromekelleher, @dcolligan Don't worry there will be a validation structure that I'm working on to make things a little simpler.  For now it would be nice to have it in one place, but feel free to implement what makes it easier for you guys.

On another note, regarding the error message I got above, I had a chance to read up on Github's documentation what might be happening and I'll just wait until it gets merged, and then I'll do a PR here afterwards against the code.

Paul

**Comment by dcolligan** [comment]
After conversations with @jeromekelleher I've decided to go with a new design.  This new code contains `import_schemas.py` which is only the functionality from `process_schemas.py` that compiles the schemas into python objects, and none of the code that does the writing of the `_protocol_schemas.py` file.  Writing and verifying `_protocol_schemas.py` will be the domain of the server repo code.  This design features minimal code duplication across both repos.

**Comment by diekhans** [comment]
At one point, I create 
  https://github.com/ga4gh/server/issues/81
  generalize python code generation from schema #81

assigned it to myself and never got back to it.

The thinking was that the class creator should be a program
independent of checking out a tree, etc.  Avro file input, class
output.  Worthy of submitting to Avro project on day.

File: tests/import_schemas.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        """
+        assert self.isSearchResponse()
+        names = [field.name for field in self.getFields()]
+        # We assume that there are exactly two fields in every
+        # SearchResponse: nextPageToken and the value list.
+        names.remove('nextPageToken')
+        assert len(names) == 1
+        return names[0]
+
+    def getEmbeddedTypes(self):
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: Shouldn't this be `checkEmbeddedTypes` or something now? We're not really expecting to return anything now, as we're just checking the schemas.

File: tests/import_schemas.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        return re.search('Search.+Request', self.name) is not None
+
+    def isSearchResponse(self):
+        """
+        Returns True if the class we are converting is a subclass of
+        SearchResponse, and False otherwise.
+        """
+        return re.search('Search.+Response', self.name) is not None
+
+    def getValueListName(self):
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: Do we use this method, or is it just a checker now?

File: tests/import_schemas.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        # SearchResponse: nextPageToken and the value list.
+        names.remove('nextPageToken')
+        assert len(names) == 1
+        return names[0]
+
+    def getEmbeddedTypes(self):
+        """
+        Returns the set of embedded types in this class.
+        """
+        # TODO need to clarify how we operate on Unions here. The current
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: We should remove this comment, and address it. What do we assume about the structure of unions in v0.6?

File: tests/import_schemas.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                        if isinstance(t1, avro.schema.RecordSchema):
+                            ret.append((field.name, t1.name))
+                    else:
+                        msg = "Schema union assumptions violated: {}.{}"
+                        raise Exception(msg.format(self.name, field.name))
+        return ret
+
+
+class SchemaGenerator(object):
+    """
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: This class name isn't appropriate now, the comment is incorrect.  SchemaChecker or SchemaVerifier?

File: tests/import_schemas.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        self.schemaDir = schemaDir
+        self.verbosity = verbosity
+        self.classes = []
+        for avscFile in glob.glob(os.path.join(self.schemaDir, "*.avsc")):
+            self.classes.append(SchemaClass(avscFile))
+        self.requestClassNames = [
+            cls.name for cls in self.classes if cls.isSearchRequest()]
+        self.responseClassNames = [
+            cls.name for cls in self.classes if cls.isSearchResponse()]
+        self.postSignatures = []
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: Are we using postSignatures? We should probably keep it so we can do some checking on these. Maybe insert a comment here to say we should add some CI on the form of the post signatures?

File: tests/import_schemas.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    def download(self, url, destination):
+        """
+        Downloads the specified url and saves the result to the specified
+        file.
+        """
+        fileDownloader = utils.FileDownloader(url, destination)
+        fileDownloader.download()
+
+    def convertAvro(self, avdlFile):
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: Do we need to do all of this? Can we assume that the files are already in avsc form, and skip the downloading the Java jar, running it etc?

File: tests/import_schemas.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+    def setupAvroJar(self):
+        if self.avroJarPath is not None:
+            self.avroJar = os.path.abspath(self.avroJarPath)
+        else:
+            url = "http://www.carfab.com/apachesoftware/avro/stable/java/"\
+                "avro-tools-1.7.7.jar"
+            self.download(url, self.avroJar)
+
+    def getSchemaFromGitHub(self):
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: I don't think we need this: travis has already cloned the repo.

**Comment by jeromekelleher** [comment]
Looks good @dcolligan, thanks. I think this is definitely the way to go. However, I think we can simplify what we're doing here a great deal. There's still a lot of stuff left over from the context of generating Python code which we should remove. I think the majority of this is just deleting things though.

File: tests/import_schemas.py (hunk 6 above)
Comment (adamnovak) [inline tests/import_schemas.py]: I don't think we can; if it's not done here, converting to avsc will have to be done in the testing script.

**Comment by adamnovak** [comment]
I agree with Jerome. It needs a pass through of renaming things.

File: tests/import_schemas.py (hunk 4 above)
Comment (dcolligan) [inline tests/import_schemas.py]: I think the name and comment are fine.  SchemaGenerator "Transforms schemas into Python objects" -- they just happen to be python objects that aren't written out to any file, as they were previously.

**Comment by dcolligan** [comment]
@jeromekelleher pushed a new commit addressing your concerns

File: tests/import_schemas.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
+"""
+A script to generate the schemas for the GA4GH protocol
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: This comment is wrong...

File: tests/import_schemas.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    def isSearchResponse(self):
+        """
+        Returns True if the class we are converting is a subclass of
+        SearchResponse, and False otherwise.
+        """
+        return re.search('Search.+Response', self.name) is not None
+
+    def checkEmbeddedTypes(self):
+        """
+        Returns the set of embedded types in this class.
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: Comments don't agree with the name here. Do we _use_ the returned value ever? If not, why build the list? People who have never seen our server repo are going to come in and try to understand this script. It's going to look very confusing and weird if we don't make it more focused on what it's supposed to be doing: *checking* the schemas, not generating Python code.

File: tests/import_schemas.py (hunk 4 above)
Comment (jeromekelleher) [inline tests/import_schemas.py]: I disagree. "Class that generates a schema in Python code from Avro definitions.". This is not what it does. A `SchemaGenerator` is something that generates a schema. We don't generate schemas here.

File: tests/import_schemas.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+class SchemaProcessor(object):
+    """
+    Transforms schemas into Python objects
+    """
+    def __init__(self, args):
+        self.version = args.version
+        self.verbosity = args.verbose
+        self.tmpDir = tempfile.mkdtemp(prefix="ga4gh_")
+        self.sourceTar = os.path.join(self.tmpDir, "schemas.tar.gz")
```
Comment (jeromekelleher) [inline tests/import_schemas.py]: Do we need to touch tarballs now?

**Comment by jeromekelleher** [comment]
@dcolligan I think that if we want to make this script the basis of the CI testing on this repo, we need to make it a lot simpler and more focused. There is too much stuff here which doesn't do anything and doesn't make any sense.

**Comment by dcolligan** [comment]
@jeromekelleher okay, gave it another pass.  let me know what you think.

**Comment by delagoya** [comment]
Build is failing. @dcolligan can you take a look?
```
======================================================================
ERROR: testSchemaProperties (test_protocol.TestValidateSchemas)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ga4gh/schemas/tests/test_protocol.py", line 48, in testSchemaProperties
    self._checkProperties(schemaClass)
  File "/home/travis/build/ga4gh/schemas/tests/test_protocol.py", line 63, in _checkProperties
    schemaClass.name, field.name))
Exception: Schema union assumptions violated: ReadAlignment.duplicateFragment
-------------------- >> begin captured logging << --------------------
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): www.carfab.com
requests.packages.urllib3.connectionpool: DEBUG: "GET /apachesoftware/avro/stable/java/avro-tools-1.7.7.jar HTTP/1.1" 200 12527927
--------------------- >> end captured logging << ---------------------
----------------------------------------------------------------------
Ran 2 tests in 61.807s
```

**Comment by dcolligan** [comment]
@delagoya yes, this is expected behavior to fail in this way.  Once #292 is merged, I will rebase this branch on top of master.

**Comment by jeromekelleher** [comment]
Much better, thanks @dcolligan! This is a great base for us to start adding more checks on the stuff in the schema.

+1.

**Comment by dcolligan** [comment]
Ok, pushed a new update and the tests pass.  We should get this merged quickly before another merge causes a break in the (yet-unmerged) CI.

Note: I also fixed some minor encoding issues in `wip/matchmaker.avdl` -- running CI locally caught these since they were making maven throw warnings; see example run below.

```
$ python scripts/run_tests.py
*** Running: "flake8 scripts tests"
*** Running: "nosetests -v tests"
testMaven (test_maven.TestMaven) ... FAIL
testSchemaProperties (test_protocol.TestValidateSchemas) ... ok

======================================================================
FAIL: testMaven (test_maven.TestMaven)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dannyc/ga4gh/schemas/tests/test_maven.py", line 31, in testMaven
    self.runCommandCheckWarnings(mvnInstall)
  File "/Users/dannyc/ga4gh/schemas/tests/test_maven.py", line 44, in runCommandCheckWarnings
    self.fail("Command '{}' issued warning(s)".format(cmd))
AssertionError: Command 'mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V' issued warning(s)
-------------------- >> begin captured stdout << ---------------------
Running 'mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V'
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/MatchResponse.java:[11,9] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/MatchResponse.java:[14,418] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/Matchmaker.java:[11,7602] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/Matchmaker.java:[11,8015] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/MatchRequest.java:[11,640] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/MatchRequest.java:[13,70] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/MatchRequest.java:[54,69] unmappable character for encoding UTF-
[WARNING] /Users/dannyc/ga4gh/schemas/target/generated-sources/avro/org/ga4gh/match/MatchRequest.java:[62,69] unmappable character for encoding UTF-

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 2 tests in 17.315s

FAILED (failures=1)
*** ERROR
```

File: tests/test_protocol.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import unittest
+
+import avro.schema
+
+import compile_schemas
+
+
+class TestValidateSchemas(unittest.TestCase):
+    """
+    Ensure the schemas conform to certain rules
```
Comment (adamnovak) [inline tests/test_protocol.py]: Any rules in particular? It looks like the only rule here is the nulls in unions thing. When we get more rules, do we want one big "follows the rules" test, or do we want a test for each rule?

**Comment by adamnovak** [comment]
I'm going to +1 this. Besides maybe wanting not to have a monolithic validator test (which at the moment isn't a big deal since it only has one rule to check), I can't find anything wrong with it.

File: tests/test_protocol.py (hunk 11 above)
Comment (dcolligan) [inline tests/test_protocol.py]: We can solve that in a subsequent PR.  I just want to get this framework checked in that has CI run against all subsequent PRs, and then we can argue about what tests we want within that framework and how to structure them.

**Comment by macieksmuga** [comment]
+1 from me as well.

**Comment by delagoya** [comment]
Merging.

**Pull request merged**